### ISSUE TITLE
PDF Download

### ISSUE THREAD

**Issue by WillDignazio** [donnemartin/system-design-primer#4]
I'd love to be able to read this as a PDF, can a pdf generated per-commit be added to the repo?

**Comment by donnemartin** [comment]
Hi @WillDignazio, interesting suggestion!  Not sure how to do this, if anyone has any tips, please feel free to share.

**Comment by dumindu** [comment]
@donnemartin If you convert this to a [Gitbook](https://github.com/donnemartin/system-design-primer/issues/7) , it gives interfaces to download this as pdf, epub and etc.

**Comment by FilipMiletic** [comment]
Yup, try Gitbook. Having this amazing piece in PDF would be great, and I would appreciate it a lot.

**Comment by addianto** [comment]
@donnemartin Or you can also use Travis CI to generate the PDF automatically and upload it as a [GitHub releases](https://docs.travis-ci.com/user/deployment/releases/). I can try setting up an example in my fork and make a Pull Request this weekend.

The idea is to use a utility named [pandoc](http://pandoc.org/) that can convert Markdown file to other format such as PDF. We can setup a Travis build by installing the required tool and dependencies for *pandoc* in order to read Markdown file and generate the PDF file.

**Comment by donnemartin** [comment]
@addianto interesting approach, thanks for investigating it and reporting the results.

@dumindu @FilipMiletic I'll try to free up and check out Gitbook, thanks for the suggestion.

**Comment by anishshah44** [comment]
Surely one day give contribute

**Comment by donnemartin** [comment]
@SuperAssassinNinja @anishshah44 I really appreciate the gesture!  Please just donate to your favorite charity, thanks :)

**Comment by andywb** [comment]
oh....

**Comment by addianto** [comment]
I continued my previous experiment on using pandoc to automate PDF generation. The tool still failed to generate the PDF file.

The cause of error probably because the image files are linked from external source, i.e. from Imgur. Is it possible to locally host the images in the repository? @donnemartin 

If anyone interested, you can see the build log [here](https://travis-ci.org/addianto/system-design-primer/builds/211606082).

**Comment by addianto** [comment]
Did a quick other experiment using a Python module named `m2pdf`. Although it required several dependencies on Ubuntu, it can read the README file in this repository and successfully generated the PDF file.

The result is not quite right yet. Several images and tables are rendered incorrectly in the generated PDF document.  See the example: https://imgur.com/a/SGdrK

**Comment by donnemartin** [comment]
Thanks for the update @addianto.  I think hosting images in the repo is do-able.  Good to hear that `m2pdf` is working quite well.

I'd like to explore the Gitbook path when I free up and compare the PDF result to `m2pdf`.

**Comment by elleryfamilia** [comment]
@WillDignazio I just used the Atom plugin "Markdown to PDF" and it did at Ok job. Too large a font size is the only issue I'm seeing.

**Comment by divyum** [comment]
Hi @donnemartin , thanks for making such an informative guide.
I tried experimenting with Gitbook, have created a very basic draft [here](https://divyumrastogi.gitbooks.io/system-design/content/the_system_design_primer/motivation.html).
Looking for feedback/suggestions so that I can continue developing it.

P.S. Topics shall be empty. Its just how the doc will look like, will have to add data to respective topics and/or create nested topics based on requirement.

**Comment by donnemartin** [comment]
Hi @divyum, thanks for prototyping Gitbook!  This does look easier to digest than the current single page README.

I'm not too familiar with Gitbook, it seems like I'd have to run the Gitbook version from my GitHub account so it auto-syncs with this repo?

**Comment by canhlinh** [comment]
[system-design-primer.pdf](https://github.com/donnemartin/system-design-primer/files/875541/system-design-primer.pdf)

**Comment by divyum** [comment]
@donnemartin I tried auto-sync repo, the issue is that Gitbook merges complete README into a single topic which is like a single README in Gitbook :P. The only way I see is to manually move data into different topics.

I also tried making gitbook from gitbook-cli but seems it just gives you a basic structure and not automatically makes topic for each section (see [here](https://github.com/divyum/system-design-primer/blob/gitbook/gitbook/SUMMARY.md)). We can also follow this structure and keep sections in different markdown files.

**Comment by donnemartin** [comment]
@canhlinh thanks for the PDF version!  I updated the [original post](https://github.com/donnemartin/system-design-primer/issues/4#issue-212919425) with a link to your comment.

@divyum thanks for the additional info!  I hope to free up soon to better evaluate/help with Gitbook.

**Comment by myartsev** [comment]
Would be great to have a PDF as a github "release"

**Comment by hectorv** [comment]
Similar to https://github.com/donnemartin/system-design-primer/pull/207, we could have something like this:
```
pandoc \
  --pdf-engine=xelatex \
  --toc -N \
  -V colorlinks -V urlcolor=NavyBlue -V toccolor=Red \
  -o system-design-primer.pdf \
  README.md
```
but it's not perfect, it still needs some tweaking.

**Comment by canhlinh** [comment]
I just generated the latest PDF here
[system-design-primer.pdf](https://github.com/donnemartin/system-design-primer/files/6987980/system-design-primer.pdf)

**Comment by dumindu** [comment]
If we can add these under releases, this will be more visible and accessible to all :)

**Comment by rcalixte** [comment]
I just opened a PR for this. 
A sample of the PDF that will be auto-generated once it is merged:
[system-design-primer.pdf](https://github.com/rcalixte/system-design-primer/releases/download/version-66cde46/system-design-primer.pdf)
[system-design-primer-zh.pdf](https://github.com/rcalixte/system-design-primer/releases/download/version-66cde46/system-design-primer-zh.pdf)

Feedback would be greatly appreciated. 🤞🏽 🙏🏽

**Comment by onurusluca** [comment]
@canhlinh Could you please update the pdf? Thank you!

**Comment by mohammedouahman** [comment]
Thanks for sharing this informative pdf, can  you please provide us how did you done that, because we need  other versions of different languages as you know.

**Comment by dmitri-mamrukov** [comment]
> Thanks for sharing this informative pdf, can you please provide us how did you done that, because we need other versions of different languages as you know.

You can run https://github.com/donnemartin/system-design-primer/blob/master/generate-epub.sh to generate an ePub, which can be converted to a PDF.

**Comment by arthurazs** [comment]
Latest [README.pdf](https://github.com/user-attachments/files/16414964/README.pdf)

**Comment by rcalixte** [comment]
> I just opened a PR for this.
>
> Feedback would be greatly appreciated. 🤞🏽 🙏🏽

The rebased and updated PDFs:

[system-design-primer.pdf](https://github.com/rcalixte/system-design-primer/releases/download/version-66cde46/system-design-primer.pdf)
[system-design-primer-zh.pdf](https://github.com/rcalixte/system-design-primer/releases/download/version-66cde46/system-design-primer-zh.pdf)
